audio_utils improvements

hollance opened the pull request:
# What does this PR do?

Recently the `audio_utils.py` file was added to Transformers to provide shared functions for audio processing such as STFT. This PR aims to clean up the code and make the API more robust.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

hollance committed:
silly change to allow making a PR

hollance committed:
clean up doc comments

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21998). All of your documentation changes will be reflected on that endpoint.

hollance committed:
simplify hertz_to_mel and mel_to_hertz

hollance committed:
fixup

hollance commented:
I cleaned up `hertz_to_mel` and `mel_to_hertz` a bit:

- more consistent doc comments
- both support single float inputs as well as numpy arrays
- simplified the formulas so it's not literally the same as the librosa code but also doesn't do pointless calculations

Since I think this implementation was based on librosa, we should also give them credit.

ArthurZucker reviewed (approved):
Thanks for taking care of the support for numpy arrays. As long as the whisper and CLAP extraction tests pass (they are not slow so should be good!) should be good!

hollance committed:
clean up power_to_db

hollance committed:
also add amplitude_to_db

hollance committed:
move functions

hollance committed:
clean up mel_filter_bank

hollance commented:
Changed `get_mel_filter_banks` into `mel_filter_bank`. Mostly renamed arguments and variables and cleaned up the doc comments, so that the naming is more in line with the rest of Transformers, e.g. `num_frequency_bins` instead of `nb_frequency_bins`.

hollance committed:
fixup

hollance committed:
credit librosa & torchaudio authors

hollance committed:
add unit tests

hollance committed:
tests for power_to_db and amplitude_to_db

hollance committed:
add mel_filter_bank tests

hollance committed:
rewrite STFT

hollance committed:
add convenience spectrogram function

hollance committed:
missing transpose

hollance committed:
fewer transposes

hollance committed:
add integration test to M-CTC-T

hollance committed:
frame length can be either window or FFT length

hollance committed:
rewrite stft API

hollance committed:
add preemphasis coefficient

hollance committed:
move argument

hollance committed:
add log option to spectrogram

hollance committed:
replace M-CTC-T feature extractor

hollance committed:
fix api thing

hollance committed:
replace whisper STFT

hollance committed:
replace whisper mel filters

hollance committed:
replace tvlt's stft

hollance committed:
allow alternate window names

hollance committed:
replace speecht5 stft

hollance committed:
fixup

hollance committed:
fix integration tests

hollance commented:
Replaced the hand-rolled STFT in the different models with the one from `audio_utils`:

- CLAP
- M-CTC-T
- SpeechT5
- TVLT
- Whisper

Did not do `audio_spectrogram_transformer` and `speech_to_text`. These use `ta_kaldi.fbank`, which is simple enough and faster than `audio_utils`. If we want to get completely rid of torchaudio we could also replace these.

hollance commented:
@sanchit-gandhi @ArthurZucker I think this is ready for review now. Feel free to look at this with a critical eye! 

The STFT code is currently written for ease of understanding and flexibility, not speed, although it does outperform the previous methods we were using.

ArthurZucker commented on src/transformers/models/mctct/feature_extraction_mctct.py:
```diff
@@ -183,36 +116,27 @@ def _extract_mfsc_features(self, one_waveform: np.array) -> np.ndarray:
+            hop_length=self.sample_stride,
+            fft_length=self.n_fft,
+            center=False,
+            preemphasis=self.preemphasis_coeff,
```
cool I think it is important to have these kind of params in the feature extractor config to know what it special about each one when extracting mel.

ArthurZucker commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,516 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
+    spectrogram = np.empty((num_frames, num_frequency_bins), dtype=np.complex64)
+
+    # rfft is faster than fft
+    fft_func = np.fft.rfft if onesided else np.fft.fft
```
I think `rfft` also supports batching !

hollance replied:
It does, and that's why librosa is currently faster. Even for a single input waveform they split it into batches.

ArthurZucker reviewed (approved):
👏🏻 Awesome work! The utils look super nice, great work on using this on the other models and adding tests were it was missing! 
I don't remember how many other models would benefit from this (maybe 2-3)  but can definitely be in a follow up PR! Same with batching that would bring even more impact! 
Great work here! 🚀

hollance commented on tests/utils/test_audio_utils.py:
```diff
@@ -0,0 +1,636 @@
+        spectrogram[1, 1] = 1.0
+
+        output = power_to_db(spectrogram, reference=1.0)
+        expected = np.array([[3.01029996, -3.01029996, -1.50580586], [-100.0, 0.0, -100.0]])
```
The spelling of this is very inconsistent in the codebase.

hollance commented on tests/utils/test_audio_utils.py:
I'm not sure how to do this?

ydshieh replied:
Hi! Could you run `python3 -m pytest -v tests/utils/test_audio_utils.py` on a CPU machine and see how long it takes?

ydshieh replied:
It takes 5 seconds on runners: it could be run on CircleCI.

hollance commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,516 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
+    return spectrogram
+
+
+def power_to_db(
```
Consistency with what, though? The reason I ordered them like they are is that `power_to_db` is done after you make a spectrogram.

hollance commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,516 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
+        fft_freqs (`np.ndarray` of shape `(num_frequency_bins,)`):
+            Discrete frequencies of the FFT bins in Hz.
+        filter_freqs (`np.ndarray` of shape `(num_mel_filters,)`):
+            Center points of the triangular filters to create, in Hz.
```
Changing it to "center frequencies"

hollance commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,516 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
-        it.
-        - This means that large variations in energy may not sound all that different if the sound is loud to begin
-        with. This compression operation makes the mel features match more closely what humans actually hear.
+    To create a log-mel spectrogram, call `np.log10(...)` on the output of this function. Note: This doesn't convert to
+    true decibels; use `amplitude_to_db()` or `power_to_db()` for that, depending on whether this is an amplitude or
+    power spectrogram.
```
Rephrasing it a bit so we don't need the "note".

hollance commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,516 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
-    log_spec = 10 * np.log10(np.clip(mel_spectrogram, a_min=a_min, a_max=None))
-    log_spec -= 10.0 * np.log10(np.maximum(a_min, ref))
-    if top_db is not None:
-        if top_db < 0:
-            raise ValueError("top_db must be non-negative")
-        log_spec = np.clip(log_spec, min=np.maximum(log_spec) - top_db, max=np.inf)
-    return log_spec
+    return 2 ** int(np.ceil(np.log2(window_length)))
```
Good point, I should actually call this function. ;-) It's a non-trivial calculation, so I think it's OK to have a function for it.

hollance committed:
fix doc comments

hollance committed:
remove manual FFT length calculation

sanchit-gandhi commented on src/transformers/models/speecht5/feature_extraction_speecht5.py:
```diff
@@ -160,46 +160,24 @@ def zero_mean_unit_var_norm(
+            mel_floor=self.mel_floor,
+            log_mel="log10",
+        )
+        return log_mel_spec.T
```
(nit) Is it clearer to transpose here or in the `__call__`? Just wonder whether we can have consistency across our feature extractors by always returning the non-transposed version?

hollance replied:
I've used the shapes as returned by librosa and torchaudio, which is `(bins, length)` but not every model uses it in that order. For existing models I didn't want to change the shape of tensors returned by the feature extractors.

To me it makes most sense to have the feature extractor return the spectrogram in the shape the model intends to use it. 

(BTW, right now none of the feature extractors actually documents its return values.)

sanchit-gandhi replied:
Sounds good to me!

sanchit-gandhi commented on tests/utils/test_audio_utils.py:
```diff
@@ -0,0 +1,636 @@
+# coding=utf-8
```
Very comprehensive! Wonder whether we can slim it down to 3-4 tests that have max coverage (e.g. that would tell us if we break something)? Or as Arthur suggested run it as a nightly test only (cc @ydshieh)

hollance replied:
There are a lot of combinations to test, so I wouldn't want to remove any of these tests.

hollance replied:
@ydshieh On my 2019 Intel iMac it takes 14 seconds to run these tests.

hollance replied:
It just loads the `"hf-internal-testing/librispeech_asr_dummy"` dataset (like the tests of most of the audio models do) but no checkpoints.

sanchit-gandhi reviewed (approved):
Very nice PR @hollance. The audio utils code is clear, comprehensive and easy to understand. Great to see so much feature extractor code being replaced by a simple one-function calls! Given the flexibility in the code, this will really simplify new feature extractor additions going forwards.

hollance commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,515 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
+    return mel
 
+
+def optimal_fft_length(window_length: int) -> int:
```
Good point!

hollance committed:
fix docs

hollance committed:
go away, deprecation warnings

hollance commented:
@sanchit-gandhi @ArthurZucker Are you OK with the PR in its current state? Then I can ask a core maintainer for a final review.

sanchit-gandhi commented:
Took a second look through and the changes LGTM @hollance!

sgugger commented on docs/source/en/internal/audio_utils.mdx:
```diff
@@ -23,12 +23,18 @@ Most of those are only useful if you are studying the code of the image processo
 
 [[autodoc]] audio_utils.mel_to_hertz
 
-[[autodoc]] audio_utils.get_mel_filter_banks
```
They are in the main documentation, I don't know how it gets more public than this.

sgugger commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,521 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
-    return fb
-
-
-def get_mel_filter_banks(
```
Can't rename a documented function without at least a deprecation cycle. Since those are utils, let's maybe avoid all of this and not do the rename?

sgugger commented on src/transformers/audio_utils.py:
```diff
@@ -81,279 +76,521 @@ def mel_to_hertz(mels: np.array, mel_scale: str = "htk") -> np.array:
 
-# TODO @ArthurZucker: This method does not support batching yet as we are mainly focus on inference.
-def fram_wave(waveform: np.array, hop_length: int = 160, fft_window_size: int = 400, center: bool = True):
+def window_function(
```
Same comment as before on the renaming.

ydshieh replied:
@hollance As @sgugger mentioned, we can't *just*  remove/change code without a deprecation cycle when it will break things. Even if it's bad code, it's exposed and users use them. We can't just think about the clean code without considering what our users will face.

sgugger reviewed (commented):
Thanks for working on this and cleaning those up. I'm afraid renaming documented functions without taking care is not an option however, as it is breaking the public API. So either need to revert the naming (`get_mel_filter_banks`->`mel_filter_banks` is purely cosmetic so not worth doing IMO) or leave the functions with a deprecation warning (for `fram_wave` maybe?)

hollance committed:
combine everything into spectrogram function

hollance committed:
add deprecated functions back

hollance committed:
fixup

sgugger reviewed (approved):
Thanks!

hollance commented:
If everyone's happy with it, feel free to merge (I don't have rights).
